bt-core/src/main/java/bt/service/RuntimeLifecycleBinder.java

```diff
@@ -39,42 +44,60 @@ public RuntimeLifecycleBinder() {
 
     @Override
     public void onStartup(Runnable r) {
-        bindings.get(LifecycleEvent.STARTUP).add(LifecycleBinding.bind(r).build());
+        addBinding(LifecycleEvent.STARTUP, LifecycleBinding.bind(r).build());
     }
 
     @Override
     public void onStartup(String description, Runnable r) {
-        bindings.get(LifecycleEvent.STARTUP).add(LifecycleBinding.bind(r).description(description).build());
+        addBinding(LifecycleEvent.STARTUP, LifecycleBinding.bind(r).description(description).build());
     }
 
     @Override
     public void onStartup(LifecycleBinding binding) {
-        bindings.get(LifecycleEvent.STARTUP).add(binding);
+        addBinding(LifecycleEvent.STARTUP, binding);
     }
 
     @Override
     public void onShutdown(Runnable r) {
-        bindings.get(LifecycleEvent.SHUTDOWN).add(LifecycleBinding.bind(r).async().build());
+        addBinding(LifecycleEvent.SHUTDOWN, LifecycleBinding.bind(r).async().build());
     }
 
     @Override
     public void onShutdown(String description, Runnable r) {
-        bindings.get(LifecycleEvent.SHUTDOWN).add(
+        addBinding(LifecycleEvent.SHUTDOWN,
                 LifecycleBinding.bind(r).description(description).async().build());
     }
 
     @Override
     public void onShutdown(LifecycleBinding binding) {
-        bindings.get(LifecycleEvent.SHUTDOWN).add(binding);
+        addBinding(LifecycleEvent.SHUTDOWN, binding);
     }
 
     @Override
     public void addBinding(LifecycleEvent event, LifecycleBinding binding) {
-        bindings.get(event).add(binding);
+        lock.readLock().lock();
+        try {
+            if (this.eventsRun.contains(event)) {
+                // we already ran this lifecycle event, so call the binding
+                binding.getRunnable().run();
+            } else {
+                bindings.get(event).add(binding);
+            }
+        } finally {
+            lock.readLock().unlock();
+        }
     }
 
     @Override
     public void visitBindings(LifecycleEvent event, Consumer<LifecycleBinding> consumer) {
-        bindings.get(event).forEach(consumer::accept);
+        lock.writeLock().lock();
+        try {
+            if (!this.eventsRun.add(event)) {
+                throw new IllegalStateException("LifecycleEvent " + event + " has already been run.");
```

Comment:
I wonder if this is actually worth disrupting the shutdown sequence? Maybe just log the exception and be done with it?

Reply:
Agreed. It would be a bug, but not something that we should crash things for. I will fix this.